src/target/riscv/riscv-013.c

```diff
@@ -2102,14 +2428,26 @@ int riscv013_dmi_write_u64_bits(struct target *target)
 	return info->abits + DTM_DMI_DATA_LENGTH + DTM_DMI_OP_LENGTH;
 }
 
+static int maybe_execute_fence_i(struct target *target)
+{
+	RISCV013_INFO(info);
+	RISCV_INFO(r);
+	if (info->progbufsize + r->impebreak >= 2) {
+		struct riscv_program program;
+		riscv_program_init(&program, target);
+		if (riscv_program_fence_i(&program) != ERROR_OK)
+			return ERROR_FAIL;
+		if (riscv_program_exec(&program, target) != ERROR_OK)
+			return ERROR_FAIL;
+	}
+	return ERROR_OK;
+}
+
 /* Helper Functions. */
 static int riscv013_on_step_or_resume(struct target *target, bool step)
 {
-	struct riscv_program program;
-	riscv_program_init(&program, target);
-	riscv_program_fence_i(&program);
-	if (riscv_program_exec(&program, target) != ERROR_OK)
-		LOG_ERROR("Unable to execute fence.i");
+	if (maybe_execute_fence_i(target) != ERROR_OK)
```

Comment:
this seems like an unrelated PR? Or is this related to System Bus Access for some reason?

Reply:
It's related to supporting targets that don't have any program buffer. I decided to add that support so I could easily test system bus access, by firing up a spike with no program buffer. So it's at best tangentially related. This is the feature creep I referenced in my above comment.